docs/messaging/messaging-spans.md

```diff
@@ -512,6 +513,51 @@ flowchart LR;
 | `messaging.message.id` | `"a1"` | `"a2"` | |
 | `messaging.batch.message_count` |  |  | 2 |
 
+### Batch publishing
+
+Given is a publisher that publishes a batch with two messages to a topic "Q" on
+Kafka, and two different consumers receiving one of the messages.
+
+```mermaid
+flowchart LR;
+  subgraph PRODUCER
+  direction TB
+  P[Span Publish]
+  CA[Span Create A]
+  CB[Span Create B]
+  end
+  subgraph CONSUMER1
+  direction TB
+  D1[Span Receive A]
+  end
+  subgraph CONSUMER2
+  direction TB
+  D2[Span Receive B]
+  end
+  P-- parent -->CA;
```

Comment:
I'd really prefer to change it to a more realistic and advanced example with links. 
The problem with parenting is that instrumentation assumes that messages are created and published in the same ambient context.

In reality API looks like [`send(ProducerRecord<K,V> record,
                                                        Callback callback)`](https://kafka.apache.org/10/javadoc/index.html?org/apache/kafka/clients/producer/KafkaProducer.html) and instrumentation has no idea in which context messages were created.

On Azure, we even provide buffered producer for EventHubs which publishes messages in a background thread.

As a result, in a general case, the only reasonable instrumentation point for `create` operation is message creation (in the case of Azure, we do it when a message is added to the batch).

It means that publish operation would start after creation and cannot be a parent for creation.

Long story short, I'd prefer to have links here instead of parent and also in the table below.

Reply:
it seems kafka is even more nuanced - 
https://stackoverflow.com/questions/58492689/how-to-send-batched-data-with-spring-kafka-producer.

`send` is not a send, but rather 'add-to-batch' and callback when actually sent. So this example stands as is for Kafka. But I'm a bit worried it creates a wrong impression for the general case of batch sending. Let me do a small research and let's quickly discuss on the messaging SIG on Thursday.

Reply:
Did a bit of a research on how batch publish APIs look like:

[AmazonSQS](https://docs.aws.amazon.com/AWSJavaSDK/latest/javadoc/com/amazonaws/services/sqs/AmazonSQS.html#sendMessageBatch-com.amazonaws.services.sqs.model.SendMessageBatchRequest-), [RocketMQ](https://rocketmq.apache.org/docs/4.x/producer/05message4/), Azure [EventHubs](https://javadoc.io/doc/com.azure/azure-messaging-eventhubs/latest/com/azure/messaging/eventhubs/EventHubBufferedProducerClient.html), and [EventGrid](https://javadoc.io/doc/com.azure/azure-messaging-eventgrid/latest/com/azure/messaging/eventgrid/EventGridPublisherClient.html) batch producer API take a list of messages. With such APIs it's hard to assume that publish is called in the same ambient context as message creation.

Also, thinking more about buffered client like Kafka (also Pulsar and GCP PubSub), even there parent relationships would be wrong. They buffer messages until buffer is full and then start publishing them. In this case by the time publishing starts all the ambient context in which messages were created is lost. 
![image](https://github.com/open-telemetry/semantic-conventions/assets/2347409/6cade064-5824-4e34-b345-c1c7a74c0724)

Reply:
Given this,
- Kafka or other buffering client that publishes [asynchronously](https://docs.confluent.io/kafka-clients/java/current/overview.html#asynchronous-writes) should always create messages spans in the context `send` is called, before actual publishing starts.
- Clients that publish one message [sychronously](https://docs.confluent.io/kafka-clients/java/current/overview.html#synchronous-writes) may use parent-child relationships. 
- But sync client that publish a batch should not assume messages are created in the same context in which publish is called. Therefore it's always better to create a message span when a message is created, not when it's published.

I think it would be easier to implement and visualize if we recommended links in all cases and changed

https://github.com/open-telemetry/semantic-conventions/blob/a4c60f7464eb84d581819b9d0181d9c0433f471b/docs/messaging/messaging-spans.md?plain=1#L243-L244

to something like `"Create" span SHOULD be created when message is created or passed to the client library (or other component) responsible for publishing. "Publish" span SHOULD link to the message creation contexts.`

Reply:
I adapted the example (and the wording for producer span relationships) according to our discussions in the workgroup.

@lmolkova @alevenberg @joaopgrassi Please have another look.

Reply:
Trying to prototype this, I came across some problems and questions.

Currently, we require a "publish" span to always be present on producer side. "create" spans can be present, and are especially important for batch publishing scenarios.

Furthermore, as a result of recent discussions, "publish" spans should cover the operation of the SDK sending messages to the intermediary, and should also be able to capture the duration and success of this operation.

This leads to some open questions:
1. When instrumenting an abstraction layer like JMS or MassTransit, this is hard to accomplish. The "publish" span would need to be created by messaging SDKs used under the cover, however, that's often not possible and thus makes it hard for these abstraction layers to provide consistent "publish" spans. As an alternative, "publish" spans could be created to cover `publish` calls of the API.
2. In case a customer sends a single message without creation context attached, the context of the "publish" operatino would be attached to this message. This context doesn't correspond to the API `publish` call, but possibly to some background thread taking and sending messages from a queue. With this setup, the creation context might be of limited use, as it doesn't correspond to the user context in which the message was created and passed via the API `publish` call, but to the SDK context for sending messages in the background. As an alternative, "create" spans can be created.

One problem I see with the "publish" span modelling the SDK-internal operation of publishing messages is, that it is of limited use when trying to track the origin and flow of a message. It is useful for seeing what's going inside an SDK, but I think in most cases users don't care much whether SDKs batch messages under the cover, but they rather want to correlate producer and consumer traces (and not traces relating to SDK internals).

Reply:
> When instrumenting an abstraction layer like JMS or MassTransit, this is hard to accomplish. The "publish" span would need to be created by messaging SDKs used under the cover, however, that's often not possible and thus makes it hard for these abstraction layers to provide consistent "publish" spans. As an alternative, "publish" spans could be created to cover publish calls of the API.

If I understand it right, you mean the abstraction (e.g. MassTransit) exposes a `publish` methods developers use. But under the hood, it invokes another publish from the actual messaging library that actually does the work? And in some cases, the abstraction does not "send right away" but can do all sort of other things (internally caching messages and send in a batch and etc), so the "publish" of the abstraction is not really useful/measure the real API call?

Reply:
Another problem I came across with the "publish" span modelling the SDK-internal operation of publishing is related to add instrumentation on top of SDKs that aren't instrumented.

For example, that's how a message is published when using the .NET Apache Kafka client library:
```csharp
public static void handler(DeliveryReport<Null, string>)
{
    ...
}

var message = new Message<Null, string> {
  Value = "hello world"
};

producer.Produce("my-topic", message, handler);
```

This is quite a common API pattern offered by messaging SDKs. If such an SDK doesn't offer built-in instrumentation, creating a "publish" that covers the duration until the message is published as well as the publishing status is quite involved. It would require starting a span before the call to `Produce`, decouple the span from the stack of active contexts without ending it, caching it, retrieving the right cached span during a call to the `handler`, and there end it with the status extracted from the `DeliveryReport`.

Reply:
Not great for sure. But I guess we, as conventions shouldn't worry much about or? From the point that instrumentations would then add instrumentation and things would then work, or you think this assumption is also wrong/not good?

Reply:
I'll put it on the agenda for tomorrow's workgroup meeting.

Reply:
After some research and discussions, I'm convinced that @lmolkova's suggestions are feasible and I reworded the paragraph about "Create" and "Publish" spans accordingly:

> "Create" spans MAY be created when a message is created or passed to the client
library or other component responsible for publishing.  A single "Create" span
SHOULD account only for a single message.  "Publish" spans SHOULD be created
for operations of sending or publishing a message to an intermediary. A single
"Publish" span can account for a single message, or for multiple messages (in
the case of sending messages in batches).  "Publish" spans SHOULD link to the
message creation contexts.

@lmolkova @alevenberg, Please have another look.

Reply:
@pyohannes  So giving this problem above https://github.com/open-telemetry/semantic-conventions/pull/509#discussion_r1447531471, with the new wording this still occurs IIUC. I remember that we discussed in the meeting that the process of instrumenting that is involved, but not impossible, thus we are OK with it, giving it still follows our conventions?

Reply:
@joaopgrassi Yes, we said that instrumenting in this way can be cumbersome, but is certainly possible. It was a consensus that this is acceptable.